File: backend/app/services/social_post_publish_service.py
import requests

FB_API_VERSION = "v21.0"
FB_GRAPH_URL = f"https://graph.facebook.com/{FB_API_VERSION}"


def publish_photo_to_facebook_page(page_id: str, page_access_token: str, image_path: str, caption: str) -> dict:
    """
    Uploads a local image file and publishes it as a photo post to a Facebook Page,
    with the given caption as the post's message.

    Returns a dict: {"success": bool, "external_id": str | None, "external_url": str | None, "error": str | None}
    """
    url = f"{FB_GRAPH_URL}/{page_id}/photos"

    try:
        with open(image_path, "rb") as image_file:
            files = {"source": image_file}
            data = {
                "caption": caption,
                "access_token": page_access_token,
            }
            response = requests.post(url, files=files, data=data, timeout=30)
    except FileNotFoundError:
        return {"success": False, "external_id": None, "external_url": None, "error": f"Image file not found: {image_path}"}
    except requests.RequestException as e:
        return {"success": False, "external_id": None, "external_url": None, "error": f"Network error calling Facebook: {str(e)}"}

    if not response.ok:
        try:
            error_detail = response.json().get("error", {}).get("message", response.text)
        except ValueError:
            error_detail = response.text
        return {"success": False, "external_id": None, "external_url": None, "error": error_detail}

    result = response.json()
    post_id = result.get("post_id") or result.get("id")

    external_url = None
    if post_id:
        # post_id from /photos is usually "{page_id}_{post_id}" — build a viewable link
        external_url = f"https://www.facebook.com/{post_id}"

    return {"success": True, "external_id": post_id, "external_url": external_url, "error": None}

File: frontend/lib/social-post-publish-service.ts
import { readFile } from "fs/promises";
import { existsSync } from "fs";

interface PublishResult {
  success: boolean;
  external_id: string | null;
  external_url: string | null;
  error: string | null;
}

/**
 * Publish a photo to a Facebook Page using the Graph API
 * 
 * @param pageId Facebook Page ID
 * @param pageAccessToken Page access token with publishing permissions
 * @param imagePath Absolute path to the image file on disk
 * @param caption Caption text for the post
 * @returns Promise<PublishResult> - Never throws, always returns success/error info
 */
export async function publishPhotoToFacebookPage(
  pageId: string,
  pageAccessToken: string,
  imagePath: string,
  caption: string
): Promise<PublishResult> {
  try {
    // Check if file exists
    if (!existsSync(imagePath)) {
      return {
        success: false,
        external_id: null,
        external_url: null,
        error: "Image file not found",
      };
    }

    // Read the image file from disk
    const imageBuffer = await readFile(imagePath);
    
    // Create a Blob from the buffer
    const imageBlob = new Blob([imageBuffer], { type: "image/png" });

    // Create FormData for multipart/form-data request
    const formData = new FormData();
    formData.append("source", imageBlob, "image.png");
    formData.append("caption", caption);
    formData.append("access_token", pageAccessToken);

    // Make the API call to Facebook Graph API
    const response = await fetch(`https://graph.facebook.com/v21.0/${pageId}/photos`, {
      method: "POST",
      body: formData,
    });

    if (!response.ok) {
      // Try to get error details from Facebook's response
      let errorMessage = `HTTP ${response.status}: ${response.statusText}`;
      try {
        const errorData = await response.json();
        if (errorData.error && errorData.error.message) {
          errorMessage = `Facebook API Error: ${errorData.error.message}`;
          if (errorData.error.code) {
            errorMessage += ` (Code: ${errorData.error.code})`;
          }
        }
      } catch {
        // If we can't parse the error response, use the HTTP status
      }

      return {
        success: false,
        external_id: null,
        external_url: null,
        error: errorMessage,
      };
    }

    // Parse successful response
    const data = await response.json();
    
    // Facebook returns { id: "pageId_postId", post_id: "pageId_postId" }
    const postId = data.post_id || data.id;
    
    if (!postId) {
      return {
        success: false,
        external_id: null,
        external_url: null,
        error: "Facebook API returned success but no post ID",
      };
    }

    // Build the Facebook URL
    const externalUrl = `https://www.facebook.com/${postId}`;

    return {
      success: true,
      external_id: postId,
      external_url: externalUrl,
      error: null,
    };
  } catch (error) {
    // Catch all other errors (network, file system, etc.)
    let errorMessage = "Unknown error occurred";
    
    if (error instanceof Error) {
      errorMessage = error.message;
    } else if (typeof error === "string") {
      errorMessage = error;
    }

    return {
      success: false,
      external_id: null,
      external_url: null,
      error: errorMessage,
    };
  }
}